### ISSUE TITLE
Merged definitions of member functions distinguished by non-functionally-equivalent constraints and by deduced return type

### ISSUE THREAD

**Issue by ecatmur** [llvm/llvm-project#61273]
```c++
template<int I> concept C = true;
template<typename T> struct A {
    auto f() requires C<42> { return 1; }  // #1
    auto f() requires true { return 2u; }  // #2
};
int main() {
    int (A<int>::*p)() = &A<int>::f;
    unsigned (A<int>::*q)() = &A<int>::f;
    return (A<int>().*p)() + (A<int>().*q)();
}
```
The above program is accepted and returns 4; the expected result would be 3. Inspection of the code (by O0 or by volatile) determines that a single `A<int>::f()` is emitted, corresponding to #.2. 

Very strangely, replacing `2u` by `2l` and `unsigned` by `long` causes clang to reject: "definition with same mangled name '_ZN1AIiE1fEv' as another definition". This seems more sensible, but is apparently contra the Standard says: https://github.com/cplusplus/CWG/issues/256. I have pointed out that following the Standard would require mangling constraints, which I view as a non starter since it would break ABI now and in future.

**Comment by erichkeane** [comment]
Definitely looks like it.  We don't have guidance from the Itanium ABI folks yet as far as I know (and this would be something for one of the codegen code owners to take a look at: @rjmccall @efriedma-quic @asl).

**Comment by zygoloid** [comment]
Fixed in [4b163e3](https://github.com/llvm/llvm-project/commit/4b163e343cfa54c8d55c9da73c70d58f55ea9df2).
